src/genTypeFlow.ts

```diff
@@ -36,18 +36,38 @@ function getResourceArg(resourceConfig: ResourceConfig, resourcePath: ReadonlyAr
         >`;
 }
 
+export function getNewKeyTypeFromBatchKeySetType(batchKey: string, resourceArgs: string) {
+    return `\
+        $Call<
+            ExtractArg,
+            [$PropertyType<$PropertyType<${resourceArgs}, '${batchKey}'>, 'has'>]
+        >`;
+}
+
 export function getLoaderTypeKey(resourceConfig: ResourceConfig, resourcePath: ReadonlyArray<string>) {
     // TODO: We assume that the resource accepts a single dict argument. Let's
     // make this configurable to handle resources that use seperate arguments.
     const resourceArgs = getResourceArg(resourceConfig, resourcePath);
 
-    return resourceConfig.isBatchResource
-        ? `
-        {|
-            ...$Diff<${resourceArgs}, { ${resourceConfig.batchKey}: $PropertyType<${resourceArgs}, '${resourceConfig.batchKey}'> }>,
-            ...{| ${resourceConfig.newKey}: $ElementType<$PropertyType<${resourceArgs}, '${resourceConfig.batchKey}'>, 0> |}
-        |}`
-        : resourceArgs;
+    if (resourceConfig.isBatchResource) {
+        let newKeyType = `${resourceConfig.newKey}: $ElementType<$PropertyType<${resourceArgs}, '${resourceConfig.batchKey}'>, 0>`;
+
+        if (resourceConfig.isBatchKeyASet) {
+            newKeyType = `${resourceConfig.newKey}: ${getNewKeyTypeFromBatchKeySetType(
+                resourceConfig.batchKey,
+                resourceArgs,
+            )}`;
+        }
```

Comment:
may be worth a comment explaining "batchKey could be part of an array or a set" or something to explain what's going on here in the default / branching case?